### ISSUE TITLE
hashtable weirdness in src/utils/topology_map.C

### ISSUE THREAD

**Issue by benkirk** [libMesh/libmesh#550]
I've got my buildbot machine that runs a bunch of old compilers on a CentOS 6.6 host, and it's choking on some new code in topology_map.C. I'm sorry this buildbot is not publicly visible, but here's the bit:
@roystgnr, any thoughts?  Intel 12.1+ and GCC 4.4+ are good...

GCC 3.4:

```pre
/usr/lib/gcc/x86_64-redhat-linux/3.4.6/../../../../include/c++/3.4.6/ext/hashtable.h: In member function `size_t __gnu_cxx::hashtable<_Val, _Key, _HashFcn, _ExtractKey, _EqualKey, _Alloc>::_M_bkt_num_key(const _Key&, size_t) const [with _Val = std::pair<const std::pair<libMesh::dof_id_type, libMesh::dof_id_type>, libMesh::dof_id_type>, _Key = std::pair<libMesh::dof_id_type, libMesh::dof_id_type>, _HashFcn = __gnu_cxx::hash<std::pair<libMesh::dof_id_type, libMesh::dof_id_type> >, _ExtractKey = std::_Select1st<std::pair<const std::pair<libMesh::dof_id_type, libMesh::dof_id_type>, libMesh::dof_id_type> >, _EqualKey = std::equal_to<std::pair<libMesh::dof_id_type, libMesh::dof_id_type> >, _Alloc = std::allocator<unsigned int>]':
/usr/lib/gcc/x86_64-redhat-linux/3.4.6/../../../../include/c++/3.4.6/ext/hashtable.h:508:   instantiated from `size_t __gnu_cxx::hashtable<_Val, _Key, _HashFcn, _ExtractKey, _EqualKey, _Alloc>::_M_bkt_num_key(const _Key&) const [with _Val = std::pair<const std::pair<libMesh::dof_id_type, libMesh::dof_id_type>, libMesh::dof_id_type>, _Key = std::pair<libMesh::dof_id_type, libMesh::dof_id_type>, _HashFcn = __gnu_cxx::hash<std::pair<libMesh::dof_id_type, libMesh::dof_id_type> >, _ExtractKey = std::_Select1st<std::pair<const std::pair<libMesh::dof_id_type, libMesh::dof_id_type>, libMesh::dof_id_type> >, _EqualKey = std::equal_to<std::pair<libMesh::dof_id_type, libMesh::dof_id_type> >, _Alloc = std::allocator<unsigned int>]'
/usr/lib/gcc/x86_64-redhat-linux/3.4.6/../../../../include/c++/3.4.6/ext/hashtable.h:447:   instantiated from `__gnu_cxx::_Hashtable_iterator<_Val, _Key, _HashFcn, _ExtractKey, _EqualKey, _Alloc> __gnu_cxx::hashtable<_Val, _Key, _HashFcn, _ExtractKey, _EqualKey, _Alloc>::find(const _Key&) [with _Val = std::pair<const std::pair<libMesh::dof_id_type, libMesh::dof_id_type>, libMesh::dof_id_type>, _Key = std::pair<libMesh::dof_id_type, libMesh::dof_id_type>, _HashFcn = __gnu_cxx::hash<std::pair<libMesh::dof_id_type, libMesh::dof_id_type> >, _ExtractKey = std::_Select1st<std::pair<const std::pair<libMesh::dof_id_type, libMesh::dof_id_type>, libMesh::dof_id_type> >, _EqualKey = std::equal_to<std::pair<libMesh::dof_id_type, libMesh::dof_id_type> >, _Alloc = std::allocator<unsigned int>]'
/usr/lib/gcc/x86_64-redhat-linux/3.4.6/../../../../include/c++/3.4.6/ext/hash_map:176:   instantiated from `typename __gnu_cxx::hashtable<std::pair<const _Key, _Tp>, _Key, _HashFcn, std::_Select1st<std::pair<const _Key, _Tp> >, _EqualKey, _Alloc>::iterator __gnu_cxx::hash_map<_Key, _Tp, _HashFcn, _EqualKey, _Alloc>::find(const typename __gnu_cxx::hashtable<std::pair<const _Key, _Tp>, _Key, _HashFcn, std::_Select1st<std::pair<const _Key, _Tp> >, _EqualKey, _Alloc>::key_type&) [with _Key = std::pair<libMesh::dof_id_type, libMesh::dof_id_type>, _Tp = libMesh::dof_id_type, _HashFcn = __gnu_cxx::hash<std::pair<libMesh::dof_id_type, libMesh::dof_id_type> >, _EqualKey = std::equal_to<std::pair<libMesh::dof_id_type, libMesh::dof_id_type> >, _Alloc = std::allocator<unsigned int>]'
src/utils/topology_map.C:73:   instantiated from here
/usr/lib/gcc/x86_64-redhat-linux/3.4.6/../../../../include/c++/3.4.6/ext/hashtable.h:518: error: no match for call to `(const __gnu_cxx::hash<std::pair<libMesh::dof_id_type, libMesh::dof_id_type> >) (const std::pair<libMesh::dof_id_type, libMesh::dof_id_type>&)'
```

Intel 11.1:
```pre
/usr/include/c++/4.4.7/backward/backward_warning.h(28): warning #1224: #warning directive: This file includes at least one deprecated or antiquated header which   may be removed without further notice at a future date. Please use a   non-deprecated interface with equivalent functionality instead. For a   listing of replacement headers and interfaces, consult the file   backward_warning.h. To disable this warning use -Wno-deprecated.
  #warning \
   ^

/usr/include/c++/4.4.7/backward/backward_warning.h(28): warning #1224: #warning directive: This file includes at least one deprecated or antiquated header which   may be removed without further notice at a future date. Please use a   non-deprecated interface with equivalent functionality instead. For a   listing of replacement headers and interfaces, consult the file   backward_warning.h. To disable this warning use -Wno-deprecated.
  #warning \
   ^

/usr/include/c++/4.4.7/backward/hashtable.h(590): error: call of an object of a class type without appropriate operator() or conversion functions to pointer-to-function type
        { return _M_hash(__key) % __n; }
                 ^
          detected during:
            instantiation of "size_t={unsigned long} __gnu_cxx::hashtable<_Val, _Key, _HashFcn, _ExtractKey, _EqualKey, _Alloc>::_M_bkt_num_key(const _Key &, size_t={unsigned long}) const [with _Val=std::pair<const std::pair<unsigned int, unsigned int>, libMesh::dof_id_type={uint32_t={unsigned int}}>, _Key=std::pair<unsigned int, unsigned int>, _HashFcn=__gnu_cxx::hash<std::pair<unsigned int, unsigned int>>, _ExtractKey=std::_Select1st<std::pair<const std::pair<unsigned int, unsigned int>,
                      libMesh::dof_id_type={uint32_t={unsigned int}}>>, _EqualKey=std::equal_to<std::pair<unsigned int, unsigned int>>, _Alloc=std::allocator<unsigned int>]" at line 594
            instantiation of "size_t={unsigned long} __gnu_cxx::hashtable<_Val, _Key, _HashFcn, _ExtractKey, _EqualKey, _Alloc>::_M_bkt_num(const _Val &, size_t={unsigned long}) const [with _Val=std::pair<const std::pair<unsigned int, unsigned int>, libMesh::dof_id_type={uint32_t={unsigned int}}>, _Key=std::pair<unsigned int, unsigned int>, _HashFcn=__gnu_cxx::hash<std::pair<unsigned int, unsigned int>>, _ExtractKey=std::_Select1st<std::pair<const std::pair<unsigned int, unsigned int>,
                      libMesh::dof_id_type={uint32_t={unsigned int}}>>, _EqualKey=std::equal_to<std::pair<unsigned int, unsigned int>>, _Alloc=std::allocator<unsigned int>]" at line 1001
            instantiation of "void __gnu_cxx::hashtable<_Val, _Key, _HashFcn, _ExtractKey, _EqualKey, _Alloc>::resize(size_t={unsigned long}) [with _Val=std::pair<const std::pair<unsigned int, unsigned int>, libMesh::dof_id_type={uint32_t={unsigned int}}>, _Key=std::pair<unsigned int, unsigned int>, _HashFcn=__gnu_cxx::hash<std::pair<unsigned int, unsigned int>>, _ExtractKey=std::_Select1st<std::pair<const std::pair<unsigned int, unsigned int>, libMesh::dof_id_type={uint32_t={unsigned int}}>>,
                      _EqualKey=std::equal_to<std::pair<unsigned int, unsigned int>>, _Alloc=std::allocator<unsigned int>]" at line 432
            instantiation of "std::pair<__gnu_cxx::_Hashtable_iterator<_Val, _Key, _HashFcn, _ExtractKey, _EqualKey, _Alloc>, bool> __gnu_cxx::hashtable<_Val, _Key, _HashFcn, _ExtractKey, _EqualKey, _Alloc>::insert_unique(const _Val &) [with _Val=std::pair<const std::pair<unsigned int, unsigned int>, libMesh::dof_id_type={uint32_t={unsigned int}}>, _Key=std::pair<unsigned int, unsigned int>, _HashFcn=__gnu_cxx::hash<std::pair<unsigned int, unsigned int>>,
                      _ExtractKey=std::_Select1st<std::pair<const std::pair<unsigned int, unsigned int>, libMesh::dof_id_type={uint32_t={unsigned int}}>>, _EqualKey=std::equal_to<std::pair<unsigned int, unsigned int>>, _Alloc=std::allocator<unsigned int>]" at line 196 of "/usr/include/c++/4.4.7/ext/hash_map"
            instantiation of "std::pair<__gnu_cxx::hashtable<std::pair<const _Key, _Tp>, _Key, _HashFn, std::_Select1st<std::pair<const _Key, _Tp>>, _EqualKey, _Alloc>::iterator, bool> __gnu_cxx::hash_map<_Key, _Tp, _HashFn, _EqualKey, _Alloc>::insert(const __gnu_cxx::hashtable<std::pair<const _Key, _Tp>, _Key, _HashFn, std::_Select1st<std::pair<const _Key, _Tp>>, _EqualKey, _Alloc>::value_type &) [with _Key=std::pair<unsigned int, unsigned int>, _Tp=libMesh::dof_id_type={uint32_t={unsigned
                      int}}, _HashFn=__gnu_cxx::hash<std::pair<unsigned int, unsigned int>>, _EqualKey=std::equal_to<std::pair<unsigned int, unsigned int>>, _Alloc=std::allocator<unsigned int>]" at line 79 of "src/utils/topology_map.C"

compilation aborted for src/utils/topology_map.C (code 2)
make[2]: *** [src/utils/libmesh_dbg_la-topology_map.lo] Error 1
```

Intel 12.0:
```pre
/usr/include/c++/4.4.7/backward/backward_warning.h(28): warning #1224: #warning directive: This file includes at least one deprecated or antiquated header which   may be removed without further notice at a future date. Please use a   non-deprecated interface with equivalent functionality instead. For a   listing of replacement headers and interfaces, consult the file   backward_warning.h. To disable this warning use -Wno-deprecated.
  #warning \
   ^

/usr/include/c++/4.4.7/backward/hashtable.h(590): error: call of an object of a class type without appropriate operator() or conversion functions to pointer-to-function type
        { return _M_hash(__key) % __n; }
                 ^
          detected during:
            instantiation of "size_t={unsigned long} __gnu_cxx::hashtable<_Val, _Key, _HashFcn, _ExtractKey, _EqualKey, _Alloc>::_M_bkt_num_key(const _Key &, size_t={unsigned long}) const [with _Val=std::pair<const std::pair<unsigned int, unsigned int>, libMesh::dof_id_type={uint32_t={unsigned int}}>, _Key=std::pair<unsigned int, unsigned int>, _HashFcn=__gnu_cxx::hash<std::pair<unsigned int, unsigned int>>, _ExtractKey=std::_Select1st<std::pair<const std::pair<unsigned int, unsigned int>,
                      libMesh::dof_id_type={uint32_t={unsigned int}}>>, _EqualKey=std::equal_to<std::pair<unsigned int, unsigned int>>, _Alloc=std::allocator<unsigned int>]" at line 594
            instantiation of "size_t={unsigned long} __gnu_cxx::hashtable<_Val, _Key, _HashFcn, _ExtractKey, _EqualKey, _Alloc>::_M_bkt_num(const _Val &, size_t={unsigned long}) const [with _Val=std::pair<const std::pair<unsigned int, unsigned int>, libMesh::dof_id_type={uint32_t={unsigned int}}>, _Key=std::pair<unsigned int, unsigned int>, _HashFcn=__gnu_cxx::hash<std::pair<unsigned int, unsigned int>>, _ExtractKey=std::_Select1st<std::pair<const std::pair<unsigned int, unsigned int>,
                      libMesh::dof_id_type={uint32_t={unsigned int}}>>, _EqualKey=std::equal_to<std::pair<unsigned int, unsigned int>>, _Alloc=std::allocator<unsigned int>]" at line 1001
            instantiation of "void __gnu_cxx::hashtable<_Val, _Key, _HashFcn, _ExtractKey, _EqualKey, _Alloc>::resize(size_t={unsigned long}) [with _Val=std::pair<const std::pair<unsigned int, unsigned int>, libMesh::dof_id_type={uint32_t={unsigned int}}>, _Key=std::pair<unsigned int, unsigned int>, _HashFcn=__gnu_cxx::hash<std::pair<unsigned int, unsigned int>>, _ExtractKey=std::_Select1st<std::pair<const std::pair<unsigned int, unsigned int>, libMesh::dof_id_type={uint32_t={unsigned int}}>>,
                      _EqualKey=std::equal_to<std::pair<unsigned int, unsigned int>>, _Alloc=std::allocator<unsigned int>]" at line 432
            instantiation of "std::pair<__gnu_cxx::_Hashtable_iterator<_Val, _Key, _HashFcn, _ExtractKey, _EqualKey, _Alloc>, bool> __gnu_cxx::hashtable<_Val, _Key, _HashFcn, _ExtractKey, _EqualKey, _Alloc>::insert_unique(const _Val &) [with _Val=std::pair<const std::pair<unsigned int, unsigned int>, libMesh::dof_id_type={uint32_t={unsigned int}}>, _Key=std::pair<unsigned int, unsigned int>, _HashFcn=__gnu_cxx::hash<std::pair<unsigned int, unsigned int>>,
                      _ExtractKey=std::_Select1st<std::pair<const std::pair<unsigned int, unsigned int>, libMesh::dof_id_type={uint32_t={unsigned int}}>>, _EqualKey=std::equal_to<std::pair<unsigned int, unsigned int>>, _Alloc=std::allocator<unsigned int>]" at line 196 of "/usr/include/c++/4.4.7/ext/hash_map"
            instantiation of "std::pair<__gnu_cxx::hashtable<std::pair<const _Key, _Tp>, _Key, _HashFn, std::_Select1st<std::pair<const _Key, _Tp>>, _EqualKey, _Alloc>::iterator, bool> __gnu_cxx::hash_map<_Key, _Tp, _HashFn, _EqualKey, _Alloc>::insert(const __gnu_cxx::hashtable<std::pair<const _Key, _Tp>, _Key, _HashFn, std::_Select1st<std::pair<const _Key, _Tp>>, _EqualKey, _Alloc>::value_type &) [with _Key=std::pair<unsigned int, unsigned int>, _Tp=libMesh::dof_id_type={uint32_t={unsigned
                      int}}, _HashFn=__gnu_cxx::hash<std::pair<unsigned int, unsigned int>>, _EqualKey=std::equal_to<std::pair<unsigned int, unsigned int>>, _Alloc=std::allocator<unsigned int>]" at line 79 of "src/utils/topology_map.C"

compilation aborted for src/utils/topology_map.C (code 2)
make[2]: *** [src/utils/libmesh_dbg_la-topology_map.lo] Error 1
```

**Comment by roystgnr** [comment]
At first glance:

It looks like @jwpeterson's fix didn't handle the __gnu_cxx::hash_map
case?

Try changing
#if defined(LIBMESH_HAVE_STD_UNORDERED_MAP) || defined(LIBMESH_HAVE_TR1_UNORDERED_MAP)
to
#if defined(LIBMESH_HAVE_STD_UNORDERED_MAP) || defined(LIBMESH_HAVE_TR1_UNORDERED_MAP) || defined(LIBMESH_HAVE_EXT_HASH_MAP)

in topology_map.h, and see if that catches it?

Thanks,
---
Roy

**Comment by jwpeterson** [comment]
> It looks like @jwpeterson's fix didn't handle the __gnu_cxx::hash_map
> case?

My fix went into serial_mesh.C, maybe the same fix can go into topology_map.C?

**Comment by jwpeterson** [comment]
The GCC error should be fixed by adding `LIBMESH_DEFINE_HASH_POINTERS` to topology_map.h, I think.

**Comment by roystgnr** [comment]
I don't think it's a lack of hash<T*> that's killing us here; the error messages looked like the problem was a lack of hash<pair<T,U> >

**Comment by jwpeterson** [comment]
> The GCC error

Sorry, meant Intel.

**Comment by jwpeterson** [comment]
> the problem was a lack of hash<pair<T,U> >

Ah, yeah, line 73.  Maybe we can fix it the same way as the pointer case, though, by adding a specialization?  It's possible I didn't see this on our cluster because I didn't build in debug mode...

**Comment by roystgnr** [comment]
Sorry I let this slip my mind for so long.  Would it be sufficient to change
```
#if defined(LIBMESH_HAVE_STD_UNORDERED_MAP) || defined(LIBMESH_HAVE_TR1_UNORDERED_MAP)
```
to
```
#if defined(LIBMESH_HAVE_STD_UNORDERED_MAP) || defined(LIBMESH_HAVE_TR1_UNORDERED_MAP) || defined(LIBMESH_HAVE_EXT_HASH_MAP) || defined(LIBMESH_HAVE_HASH_MAP)
```
?

**Comment by roystgnr** [comment]
I've just added the same problem again in coupling_matrix.h, haven't I?  At least I can replicate it (with gcc 3.4) now; I'll see if I can fix it.

**Comment by roystgnr** [comment]
This should have been fixed by #597; reopen the ticket if it's still a problem for you (or with the compilers I couldn't test against).
